src/libraries/Microsoft.PowerFx.Interpreter/EvalVisitor.cs

```diff
@@ -267,16 +269,35 @@ public override async ValueTask<FormulaValue> Visit(CallNode node, EvalVisitorCo
             {                
                 return await asyncConnectorTexlFunction.InvokeAsync(args, _services, _cancellationToken).ConfigureAwait(false);
             }
-            else if (func is UserDefinedTexlFunction udtf)
-            {
-                // $$$ Should add _runtimeConfig
-                result = await udtf.InvokeAsync(args, _cancellationToken, context.StackDepthCounter.Increment()).ConfigureAwait(false);
-            }
             else if (func is CustomTexlFunction customTexlFunc)
             {
                 // If custom function throws an exception, don't catch it - let it propagate up to the host.
                 result = await customTexlFunc.InvokeAsync(FunctionServices, args, _cancellationToken).ConfigureAwait(false);
             }
+            else if (func is UserDefinedFunction userDefinedFunc)
+            {
+                var recordValue = userDefinedFunc.GetRuntimeRecordValue(args, _cancellationToken);
+
+                var symbolsTableRecord = ReadOnlySymbolTable.NewFromRecord(recordValue.Type);
+                var symbolValuesRecord = SymbolValues.NewFromRecord(symbolsTableRecord, recordValue);
+
+                var symbolsTable = ReadOnlySymbolTable.Compose(_symbolValues.SymbolTable, symbolsTableRecord);
+                var symbolValues = ReadOnlySymbolValues.Compose(_symbolValues, symbolValuesRecord);
+
+                userDefinedFunc.SetFormulaValues(args);

```

Comment:
>userDefinedFunc.SetFormulaValues(args); [](http://example.com/codeflow?start=16&length=39)

This could be a problem - we should treat userDefinedFunc as immutable; consider if the same function is called  multiple times:

Foo(1,2, Foo(3,4,5));

Reply:
Removed. Added stack to EvalVisitor.